### PR TITLE
Inject dependencies object rather than using global state

### PR DESCRIPTION
**Pull request by bizz84**
@yjbanov This PR removes global state in favour of a single `Dependencies` object passed to all classes. Perf tests show a further small improvement.

### CODE REVIEW

File: lib/timer_page.dart [hunk 1]
```diff
@@ -220,6 +214,6 @@ class HundredsState extends State<Hundreds> {
   @override
   Widget build(BuildContext context) {
     String minutesStr = (hundreds % 100).toString().padLeft(2, '0');
-    return new Text(minutesStr, style: timerTextStyle, textAlign: TextAlign.start);
```
Comment (bizz84) [inline lib/timer_page.dart]: Using `TextAlign.start`, `TextAlign.end` made the text non-horizontally centred, so I removed it.

File: lib/timer_page.dart [hunk 2]
```diff
@@ -113,62 +130,40 @@ class TimerTextState extends State<TimerText> {
...
-
-  ElapsedTime({
-    this.hundreds,
-    this.seconds,
-    this.minutes,
-  });
-}
-
-final List<ValueChanged<ElapsedTime>> timerListeners = <ValueChanged<ElapsedTime>>[];
-const TextStyle timerTextStyle = const TextStyle(fontSize: 60.0, fontFamily: "Open Sans");
```
Comment (bizz84) [inline lib/timer_page.dart:152]: Removed these in favour of `Dependencies` class.

File: lib/timer_page.dart [hunk 3]
```diff
@@ -113,62 +130,40 @@ class TimerTextState extends State<TimerText> {
...
-          flex: 1,
-          child: new RepaintBoundary(
+          new RepaintBoundary(
             child: new SizedBox(
               height: 72.0,
-              child: new Hundreds(stopwatch),
+              child: new Hundreds(dependencies: dependencies),
             ),
           ),
-        ),
```
Comment (bizz84) [inline lib/timer_page.dart:133]: Removed `Expanded` widgets: layout looks the same without.

File: pubspec.lock [hunk 4]
```diff
@@ -85,11 +85,23 @@ packages:
       url: "https://pub.dartlang.org"
```
Comment (yjbanov) [inline pubspec.lock]: Traditionally in Dart the lock file is not checked into the repository and added to `.gitignore` instead. This file is automatically generated from `pubspec.yaml` (which _should_ be checked in) by `flutter packages get`.

**Review by yjbanov — APPROVED** [review]
Much nicer!

File: pubspec.lock (hunk 4 above)
Comment (bizz84) [reply pubspec.lock]: Good suggestion. Thanks

File: lib/timer_page.dart (hunk 3 above)
Comment (yjbanov) [reply lib/timer_page.dart:133]: Are you sure it's the same? The reason I used `Expanded` was to make sure that `MinutesAndSeconds` get 2/3 of the screen and `Hundreds` gets 1/3 of the screen. Without it, when you start the clock the dot jumps around the screen a lot.

Comment (yjbanov) [reply lib/timer_page.dart:133]: Ideally though, the layout would be such that the text doesn't shift around. A monospace font could do that. Or the clock could be split further into Minutes, Seconds and Hundreds, such that each of them has a fixed width.

Comment (bizz84) [reply lib/timer_page.dart:133]: The numerals for the font I'm using (Open Sans) all have the same width. When I test it all text stays in place.

Comment (yjbanov) [reply lib/timer_page.dart:133]: The width of digits changes when I run the app on iOS, but it's fixed when running on Android. I think what's happening is that "Open Sans" is mapped to different system fonts (the two OSes have different sets of preinstalled fonts). On Android it happens to map to one with fixed-width digits. Not sure what it maps to on iOS. To get consistent behavior you might want to bundle your own font: https://flutter.io/custom-fonts/

Comment (bizz84) [reply lib/timer_page.dart:133]: Ok. I'll look at fonts separately.

**Pull request merged**